src/socket.rs

```diff
@@ -110,6 +75,54 @@ pub struct Socket {
 }
 
 impl Socket {
+    /// Creates a new socket ready to be configured.
+    ///
+    /// This function corresponds to `socket(2)` on Unix and `WSASocketW` on Windows.
+    ///
+    /// # Notes
+    ///
+    /// On Unix-like systems, the close-on-exec flag is set on the new socket.
+    /// Additionally, on macOS X and iOS the `SOCK_NOSIGPIPE` option is set.
+    /// On Windows, the socket handle is made non-inheritable.
+    ///
+    /// If possible, the close-on-exec flag will be set atomically when the socket is created.
+    /// However, on some platforms the flag is to be set with a separate syscall,
+    /// which leaves a small window for race conditions.
+    ///
+    /// Use [`Self::new_raw`] if you don't want these flags and options to be set.
+    pub fn new(domain: Domain, ty: Type, protocol: Option<Protocol>) -> io::Result<Socket> {
```

Comment:
Please use the same structure I used here: https://github.com/rust-lang/socket2-rs/pull/125/files#diff-917ff224180b881299cdcef7bc14e710509229092ea4e6599861680bc0a2a458R82-R117, comments included.